### ISSUE TITLE
Applicability of 2.5.3 Label in Name when no label is present (or just placeholder)

### ISSUE THREAD

**Issue by detlevhfischer** [w3c/wcag#855]
WCAG 2.1 evaluators here have raised the questions regarding the applicability of 2.5.3 _Label in Name_.

1. When, say, a search field has no label (`placeholder` only) which already leads to a FAIl  3.3.2 Labels or Instructions, would content then **also** fail 2.5.3? Many UA/AT combos do use the visible placeholder as fall-back accessible name (but it does not work in some AT, such as Dragon naturally speaking). And, then, of course, the palceholder text disappears after input.

2. For cases where there **is** a visible label but this is not programmatically linked (via `label for=` or _aria-labelledby_) I guess the 2.5.3 test would check if an accName is provided by some other means and if the visible text is included in that accName (again the question would be if this should include the `placeholder` fallback).

**Comment by alastc** [comment]
1. Logically if it fails 3.3.2 it shouldn't _also_ fail 2.5.3 as placeholder isn't considered valid as a label.

2. I agree with your guess, the visible text doesn't have to be the source of the accname, it just has to match or be included.

**Comment by kantoche** [comment]
I agree with @alastc statements and I'm surprised to read in the note about placeholder:

> As such, in the absence of any other nearby text string (as described in the preceding list), if an input contains placeholder text, such text may be a candidate for Label in Name. **This is supported both through the accessible name calculation (discussed later)** and from the practical sense that where a visible label is not otherwise provided, it is likely that a speech-input user may attempt to use the placeholder text value as a means of interacting with the input.

The passage I'm pointing out (in bold) suggests that the placeholder participates in the calculation of the accessible name, whereas this is not the case. Did I misunderstand this passage? If not, it should be removed to avoid this confusion.

**Comment by patrickhlauke** [comment]
> placeholder participates in the calculation of the accessible name, whereas this is not the case

I seem to recall from testing that this varies between browsers at the moment. Some - I seem to remember Chrome and Firefox, but I may be wrong - do use `placeholder` as a fallback in the absence of anything else. but that's probably more an issue that should be tackled in the accname spec.

**Comment by JAWS-test** [comment]
> This is supported both through the accessible name calculation (discussed later) 

This part of the sentence should be removed because it is wrong. And what does "discussed later" actually mean?

> from the practical sense that where a visible label is not otherwise provided, it is likely that a speech-input user may attempt to use the placeholder text value as a means of interacting with the input

I consider this part of the sentence to be meaningful. The point is that an input field might not have a visible label, or might not have a visible label that is located nearby and can be uniquely assigned to the element. If the element now has a label via aria-label or aria-labelledby and a placeholder, then placeholder (visible label) and programmatic label should match to fulfill 2.5.3

**Comment by JAWS-test** [comment]
> Some - I seem to remember Chrome and Firefox, but I may be wrong - do use placeholder as a fallback in the absence of anything else

IE 11, Chrome und Firefox use placeholder as a fallback for a label if there is no other label. This works with JAWS and NVDA. Sometimes the placeholder attribute is also output as a description if a label exists as an accessible name.

**Comment by patrickhlauke** [comment]
coming back to this (as i'm digging through my inbox)

> When, say, a search field has no label (placeholder only) which already leads to a FAIL 3.3.2 Labels or Instructions, would content then also fail 2.5.3 or is 2.5.3 'not applicable'? Many UA/AT combos do use the visible placeholder as fall-back accessible name (but it does not work in some AT, such as Dragon naturally speaking). And, then, of course, the placeholder text disappears after input.

i'm now more strongly thinking that this situation does indeed fail 2.5.3 as well.

While a field is empty, it technically passes ... the accessible name calculation does seem to allow for `placeholder` to be used, in absence of anything else, as accessible name, and (with some variations it seems) user agents support this too. so, when the field is empty, the visible placeholder text acts as the label, and the accessible name matches this (provided there's nothing else giving this an accessible name).

However, when the field is filled in, and the placholder text isn't visible, the only "label" as such is the filled in text, which will differ from the accessible name. Which fails 2.5.3.

So, similar to the case of 3.3.2, fields that only use `placeholder` and have no other visible label technically pass when empty, but fail as soon as the placeholder text itself is not shown anymore.

**Comment by awkawk** [comment]
I think that if the placeholder doesn't count as a label due to it being temporary, then the start of 2.5.3 kicks in:

For user interface components **with labels that include text or images of text**

So as a result, if there isn't a label, then the field is not a "user interface components with labels", so 2.5.3 is N/A. As soon as you fix the label, then 2.5.3 would need to be re-evaluated.

**Comment by patrickhlauke** [comment]
yeah, probably now leaning more towards the N/A part (while the input with just placeholder visible may technically pass, once it's filled in and disappears yeah, formally there's nothing acting as a label so not applicable).

so i think we're agreed in general (going back to detlev's questions):

1. 2.5.3 is "not applicable"
2. yes, doesn't matter if the visible text is actually the one that technically gives the control its accessible name. what counts is the end result (i.e. something like `<p>Username</p> <input ... aria-label="Username">` is perfectly ok and passes)

**Comment by patrickhlauke** [comment]
one situation that i'd still say fails 2.5.3 is if there's no other visible label other than `placeholder`, BUT the input has been given another explicit accessible name that doesn't contain the placeholder text.

`<input ... placeholder="Visible text acting as label" aria-label="Actual accessible name">`

for the time where the placeholder text is visible and acts effectively as the label until overwritten by user input, the `placeholder` text is not contained in the `aria-label` accessible name and fails. (and yes, once filled in, it becomes not applicable, but I think the temporary failure state to start with counts here).

(i mean yes, in the end, this input will need to be fixed anyway as it fails 3.3.2, but still, for consistency...)

**Comment by awkawk** [comment]
I don't think that you can compare text that fails as a label with the accessible name and if they don't match, fail 2.5.3.

Either the control has a label that meets 3.3.2 or you don't. If it doesn't then I don't think that it is appropriate to do the comparison.

**Comment by patrickhlauke** [comment]
the placeholder text fails as a label because it disappears so isn't always there. while it is there it acts as a label. (i.e. imagine a voice user coming to the page, seeing fields with placeholder text of "Username", "Password", etc ... of course they'll interpret those as labels at that point. for all they know, those may even be floating labels that move to the side once you start typing etc). for 2.5.3 when it IS visible and doesn't match the accessible name, it fails, i'd argue.

again, either way that patterns needs to change in the markup either way, and it's unlikely to happen all that often in the real world...

**Comment by mraccess77** [comment]
The is interesting because the ACT Rules group has said that 2.5.3 doesn't apply to inputs  https://act-rules.github.io/rules/2ee8b8 (I don't agree with their interpretation)

"Inapplicable Example 2
This email text field does not need to have its visible text match the accessible name. The content of a textfield shows its value instead of its label; it does not support name from content. The label is usually adjacent to the textfield instead.

<div>E-mail</div>
<input type="email" aria-label="E-mail" value="Contact" />"

**Comment by awkawk** [comment]
@mraccess77 that is surprising, and definitely incorrect. @WilcoFiers - check out Jon's comment ^^^^

**Comment by srinivasuchakravarthula** [comment]
Yes, Placeholder fails as a label but it may pass when input field receive
focus and label jumps above the field and remains visible. Floating labels
have become practice for some time.
Thanks,

Srinivasu Chakravarthula, CPWA
Accessibility, Section 508, WCAG, GIGW
Phone: +91 99008 10881
Website: http://www.srinivasu.org | http://www.hemavasu.in
Senior Accessibility Program Manager, Watermark Inc., (
http://watermarkinsights.com)
Founder, ServeOM Inclusion (http://www.serveominclusion.com)
Individual Member, International Association of Accessibility Professionals
(IAAP) (http://www.accessibilityassociation.org)
Hon. Joint Secretary, The National Association for the Blind, Karnataka
Branch (http://nabkarnataka.org)

What's CPWA? Certified Professional in Web Accessibility by IAAP. This will
be awarded once CPACC (Certified Professional in Accessibility Core
Competencies) and WAS (Web Accessibility Specialists) are completed
successfully.
